### ISSUE TITLE
map_clone should not suggest copied() for MSRV < 1.36

### ISSUE THREAD

**Issue by djc** [rust-lang/rust-clippy#8623]
### Summary

I'm backporting some code to work on 1.32 and noticed that the map_clone lint suggests using copied() even though I've configured Clippy with MSRV = 1.32. Would be nice if this wouldn't happen.

### Lint Name

map_clone

**Comment by giraffate** [comment]
I tried the case below but can't reproduce it. Can you provide the reproduction?
```
$ cat src/main.rs
fn main() {
    let _: Option<u64> = Some(&16).map(|b| *b);
}
$ cargo -V
cargo 1.61.0-nightly (65c826642 2022-03-09)
$ cargo clippy
    Checking map_clone_msrv v0.1.0 ((snip)/map_clone_msrv)
warning: you are using an explicit closure for copying elements
 --> src/main.rs:2:26
  |
2 |     let _: Option<u64> = Some(&16).map(|b| *b);
  |                          ^^^^^^^^^^^^^^^^^^^^^ help: consider calling the dedicated `copied` method: `Some(&16).copied()`
  |
  = note: `#[warn(clippy::map_clone)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_clone

warning: `map_clone_msrv` (bin "map_clone_msrv") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.37s
$ cargo +1.32 -V
cargo 1.32.0 (8610973aa 2019-01-02)
$ cargo +1.32 clippy
    Checking map_clone_msrv v0.1.0 ((snip)/map_clone_msrv)
warning: You are using an explicit closure for cloning elements
 --> src/main.rs:2:26
  |
2 |     let _: Option<u64> = Some(&16).map(|b| *b);
  |                          ^^^^^^^^^^^^^^^^^^^^^ help: Consider calling the dedicated `cloned` method: `Some(&16).cloned()`
  |
  = note: #[warn(clippy::map_clone)] on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_clone

    Finished dev [unoptimized + debuginfo] target(s) in 0.58s
```

I also tried this.
```
$ cat clippy.toml
msrv = "1.32.0"%
$ cat src/main.rs
fn main() {
    let _: Option<u64> = Some(&16).map(|b| *b);
}
$ cargo -V
cargo 1.61.0-nightly (65c826642 2022-03-09)
$ cargo clippy
warning: you are using an explicit closure for copying elements
 --> src/main.rs:2:26
  |
2 |     let _: Option<u64> = Some(&16).map(|b| *b);
  |                          ^^^^^^^^^^^^^^^^^^^^^ help: consider calling the dedicated `cloned` method: `Some(&16).cloned()`
  |
  = note: `#[warn(clippy::map_clone)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_clone

warning: `map_clone_msrv` (bin "map_clone_msrv") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.30s
```

**Comment by djc** [comment]
So I was referring to `Iterator::copied()` rather than `Option::copied()`. See the change here:

https://github.com/djc/tz-tmp/commit/3714d2d725c891cb298efbbb65df24a65891b63d#diff-4a04259da480a6b794a2e947e4cc03eff4d1aa9330836f5b91cac68c5398193fR10

**Comment by giraffate** [comment]
Thanks! I can reproduce it.

**Comment by djc** [comment]
Great, thanks for working on this!

**Comment by kyoto7250** [comment]
I could not reproduce yet... 😢 ( I'm confused... )

I pulled this commit and running clippy, but the message was collect. (but I noticed message is wrong, so I will fix...)

```bash
~/repos/tz-tmp on  HEAD (3714d2d) [!] is 📦 v0.6.7 via 🦀 v1.60.0
❯ git diff
diff --git a/src/parser.rs b/src/parser.rs
index 24378ea..8b44cca 100644
--- a/src/parser.rs
+++ b/src/parser.rs
@@ -7,7 +7,7 @@ use crate::rule::TransitionRule;
 use crate::timezone::{LeapSecond, LocalTimeType, TimeZone, Transition};
 use crate::Error;

-#[allow(clippy::map_clone)] // MSRV: 1.36
+// #[allow(clippy::map_clone)] // MSRV: 1.36
 pub(super) fn parse(bytes: &[u8]) -> Result<TimeZone, Error> {
     let mut cursor = Cursor::new(bytes);
     let state = State::new(&mut cursor, true)?;

~/repos/tz-tmp on  HEAD (3714d2d) [!] is 📦 v0.6.7 via 🦀 v1.60.0
❯ cargo -V
cargo 1.60.0 (d1fd9fe2c 2022-03-01)

~/repos/tz-tmp on  HEAD (3714d2d) [!] is 📦 v0.6.7 via 🦀 v1.60.0
❯ cargo clippy --version
clippy 0.1.60 (7737e0b5 2022-04-04)

~/repos/tz-tmp on  HEAD (3714d2d) [!] is 📦 v0.6.7 via 🦀 v1.60.0
❯ cargo clippy
warning: you are using an explicit closure for copying elements
  --> src/parser.rs:79:26
   |
79 |     let std_walls_iter = state.std_walls.iter().map(|&i| i).chain(iter::repeat(0));
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider calling the dedicated `cloned` method: `state.std_walls.iter().cloned()`
   |
   = note: `#[warn(clippy::map_clone)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_clone

warning: you are using an explicit closure for copying elements
  --> src/parser.rs:80:26
   |
80 |     let ut_locals_iter = state.ut_locals.iter().map(|&i| i).chain(iter::repeat(0));
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider calling the dedicated `cloned` method: `state.ut_locals.iter().cloned()`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_clone

warning: `tz-rs` (lib) generated 2 warnings
    Finished dev [unoptimized + debuginfo] target(s) in 0.00s
```

@giraffate  Could you please share the reproduction code?

**Comment by kyoto7250** [comment]
I share my opinion...

I think the bug in Iterator::copied came from the `unnecessary_to_owned `rule, not from the `map_clone` rule.
 
Because the `map_clone` rule changes the message according to the msrv value, but `unnecessary_to_owned` rule does not change the message.

https://github.com/rust-lang/rust-clippy/blob/89ee6aa6e32df974ef9dbb3c825afbdce441e856/clippy_lints/src/methods/unnecessary_to_owned.rs#L195


If the original code looked like the following, this bug was likely to recur... 🤔

```bash
~/repos/tz-tmp on  HEAD (3714d2d) [!] is 📦 v0.6.7 via 🦀 v1.60.0
❯ cargo clippy
warning: unnecessary use of `to_vec`
  --> src/parser.rs:79:26
   |
79 |     let std_walls_iter = state.std_walls.to_vec().into_iter().chain(iter::repeat(0));
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use: `state.std_walls.iter().copied()`
   |
   = note: `#[warn(clippy::unnecessary_to_owned)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_to_owned

warning: unnecessary use of `to_vec`
  --> src/parser.rs:80:26
   |
80 |     let ut_locals_iter = state.ut_locals.to_vec().into_iter().chain(iter::repeat(0));
   |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use: `state.ut_locals.iter().copied()`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_to_owned

warning: `tz-rs` (lib) generated 2 warnings
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
```

**Comment by giraffate** [comment]
@kyoto7250 Sorry, I misunderstood it, I wasn't able to reproduce it. Anyway, thanks for checking it out!

@djc Sorry for my carelessness, but what do you think about this?
https://github.com/rust-lang/rust-clippy/issues/8623#issuecomment-1095431658

**Comment by kyoto7250** [comment]
@djc 

I misunderstood.

I thought this commit was a fix for map_clone that changed to copied even though msrv = 1.32.

So I had no doubts about map_clone working with msrv = 1.32.

---

My understanding is that there are two kinds of rules.

One rule is to disable the output depending on the value of msrv, and the other rule is to change the warning content depending on the value of msrv (map_clone is the latter).

There seems to be a mix of rules in clippy that make warnings go away and rules that don't (maybe, it's not explicitly stated...)

**Comment by kyoto7250** [comment]
I made an issue about this issue.
https://github.com/rust-lang/rust-clippy/issues/8693

**Comment by hkBst** [comment]
Is this fixed by https://github.com/rust-lang/rust-clippy/pull/8692?

**Comment by Twil3akine** [comment]
I tried to reproduce this on the current Clippy with `msrv = "1.32.0"`.

`clippy.toml`:

```toml
msrv = "1.32.0"
````

`src/main.rs`:

```rust
fn main() {
    let xs = vec![1_u32, 2, 3];
    let _ = xs.iter().map(|&x| x);
}
```

Result:

```text
❯ cargo clippy
    Checking clippy-msrv-test v0.1.0 (/Users/twil3akine/mein/coding-space/OSS/rust-clippy/clippy-msrv-test)
warning: you are using an explicit closure for cloning elements
 --> src/main.rs:3:13
  |
3 |     let _ = xs.iter().map(|&x| x);
  |             ^^^^^^^^^^^^^^^^^^^^^ help: consider calling the dedicated `cloned` method: `xs.iter().cloned()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_clone
  = note: `#[warn(clippy::map_clone)]` on by default

warning: useless use of `vec!`
 --> src/main.rs:2:14
  |
2 |     let xs = vec![1_u32, 2, 3];
  |              ^^^^^^^^^^^^^^^^^ help: you can use an array directly: `[1_u32, 2, 3]`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_vec
  = note: `#[warn(clippy::useless_vec)]` on by default

warning: `clippy-msrv-test` (bin "clippy-msrv-test") generated 2 warnings (run `cargo clippy --fix --bin "clippy-msrv-test" -p clippy-msrv-test -- ` to apply 2 suggestions)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.03s
```

Clippy now suggests `cloned()` instead of `copied()` for `map_clone`, so I could not reproduce the original issue.

It looks like this may have been fixed by #8692

Could this issue be closed?

**Comment by samueltardieu** [comment]
Already fixed.
